### PR TITLE
[proposal] add nodepool and uniteddeployment  crd proposal

### PR DESCRIPTION
**Pull request by kadisi**
Signed-off-by: kadisi <iamkadisi@163.com>

### Ⅰ. Describe what this PR does
 
add nodepool and uniteddeployment crd proposal

ref #171 

### Ⅴ. Special notes for reviews

@Fei-Guo @charleszheng44

### CODE REVIEW

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+// Each of subsets under the UnitedDeployment is described in Topology.
+type Topology struct {
+    // Contains the details of each subset. Each element in this array represents one subset
+    // which will be provisioned and managed by UnitedDeployment.
+    // +optional
+    Subsets []Subset `json:"subsets,omitempty"`
+}
+
+// Subset defines the detail of a subset.
+type Subset struct {
```
Comment (Fei-Guo) [inline docs/enhacement/20201211-nodepool_uniteddeployment.md:273]: Suggestion to rename Subset to"division" or "partition".

Also, how should we associate a SubSet with a node pool?

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+    // +optional
+    ReadyNode int32 `json:"readyNode"`
+
+    // Total number of not ready nodes in the pool.
+    // +optional
+    NotReadyNode int32 `json:"notReadyNode"`
+
+    // The list of nodes' names in the pool
+    // +optional
+    EdgeNodes map[string]bool `json:"edgeNodes,omitempty"`
```
[the remark this answers is not in the document]
Comment (kadisi) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md:99]: ok

[the remark this answers is not in the document]
Comment (rambohe-ch) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md:99]: Why use map as the type for EdgeNodes field, how about []stirng slice

File: docs/enhacement/20201211-nodepool_uniteddeployment.md (hunk 1 above)
Comment (rambohe-ch) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md:273]: > Suggestion to rename Subset to"division" or "partition" or just "pool".
> 
> Also, how should we associate a SubSet with a node pool?
pool +1

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+
+// UnitedDeploymentSpec defines the desired state of UnitedDeployment.
+type UnitedDeploymentSpec struct {
+    // Selector is a label query over pods that should match the replica count.
+    // It must match the pod template's labels.
+    Selector *metav1.LabelSelector `json:"selector"`
+
+    // Template describes the subset that will be created.
+    // +optional
+    Template SubsetTemplate `json:"template,omitempty"`
```
[the remark this answers is not in the document]
Comment (kadisi) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md:173]: good idea

File: docs/enhacement/20201211-nodepool_uniteddeployment.md (hunk 1 above)
Comment (kadisi) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md:273]: ok

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+    // subset workload name prefix in the format '<deployment-name>-<subset-name>-'.
+    // Name should be unique between all of the subsets under one UnitedDeployment.
+    // Name is NodePool Name
+    Name string `json:"name"`
+
+    // Indicates the number of the pod to be created under this subset.
+    // +optional
+    Replicas *int32 `json:"replicas,omitempty"`
+
+    // Specification of the desired behavior of the pod.
```
[the remark this answers is not in the document]
Comment (kadisi) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md:284]: The purpose of this is to enable users to perform patch operations on workload based on a single node pool
For example： user might only want to modify a image version in a pool

Can we define it as follows：
`Patches string `json:"patches,omitempty"``

If user defines this patch, it is equivalent to implementing kubectl patch operation。

[the remark this answers is not in the document]
Comment (kadisi) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md:284]: The purpose of this is to enable users to perform patch operations on workload based on a single node pool
For example： user might only want to modify a image version in a pool

Can we define it as follows：
`Patches string `json:"patches,omitempty"``

If user defines this patch, it is equivalent to implementing kubectl patch operation。

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+         * [UnitedDeployment API](#uniteddeployment-api)
+      * [Implementation History](#implementation-history)
+
+
+# Proposal about nodepool and uniteddeployment  
+
+## Glossary
+
+NodePool:
+    A node pool is a group of nodes within a same region in the edge computing scenario.
```
Comment (rambohe-ch) [inline docs/enhacement/20201211-nodepool_uniteddeployment.md:30]: Maybe we'd better add this requirement for NodePool: The nodes in the NodePool are ideally accessible to each other through the Intranet.

Comment (kadisi) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md:30]: > Maybe we'd better add this requirement for NodePool: The nodes in the NodePool are ideally accessible to each other through the Intranet.

great

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+	// +optional
+	NodeSelectorTerm corev1.NodeSelectorTerm `json:"nodeSelectorTerm,omitempty"`
+
+	// Indicates the tolerations the pods under this pool have.
+	// A pool's tolerations is not allowed to be updated.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// Indicates the number of the pod to be created under this pool.
+	// +optional
```
Comment (Fei-Guo) [inline docs/enhacement/20201211-nodepool_uniteddeployment.md:275]: It looks that the Replica field needs to be a required field.

**Review by Fei-Guo — COMMENTED** [review]
Overall, LG. A Nit

File: docs/enhacement/20201211-nodepool_uniteddeployment.md (hunk 6 above)
Comment (kadisi) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md]: done

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+    // If specified, the Taints will be added to all nodes.
+    // +optional
+    Taints []v1.Taint `json:"taints,omitempty"`
+}
+
+// NodePoolStatus defines the observed state of NodePool
+type NodePoolStatus struct {
+    // Total number of ready nodes in the pool.
+    // +optional
+    ReadyNode int32 `json:"readyNode"`
```
Comment (Walnux) [inline docs/enhacement/20201211-nodepool_uniteddeployment.md]: Could we change it to
ReadyNodeNum int32 `json:"readyNodeNum"`

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+
+    // Total number of not ready nodes in the pool.
+    // +optional
+    NotReadyNode int32 `json:"notReadyNode"`
```
Comment (Walnux) [inline docs/enhacement/20201211-nodepool_uniteddeployment.md]: Could we change it to
unreadyNodeNum int32 `json:"unreadyNodeNum"

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+    Items           []NodePool `json:"items"`
+}
+```
+
+
+### UnitedDeployment API
+
+``` go
+
+// UnitedDeploymentConditionType indicates valid conditions type of a UnitedDeployment.
```
Comment (Walnux) [inline docs/enhacement/20201211-nodepool_uniteddeployment.md]: should we change "conditions type" to "condition type"?

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 10]
```diff
@@ -0,0 +1,374 @@
+---
+title: Proposal about nodepool and uniteddeployment
```
[the remark this answers is not in the document]
Comment (kadisi) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md]: ok

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+
+# Proposal about nodepool and uniteddeployment  
+
+## Glossary
+
+NodePool:
+    A node pool is a group of nodes within a same region in the edge computing scenario. 
+
+UnitedDeployment:
+    UnitedDeployment provides a new way to manage pods in multi-nodepools by using multiple workloads. it provides an alternative to achieve high availability in a cluster that consists of multiple nodepool - that is, managing multiple homogeneous workloads, and each workload is dedicated to a single Subset. Pod distribution across nodepools is determined by the replica number of each workload. Since each Subset is associated with a workload, UnitedDeployment can support finer-grained rollout and deployment strategies.
```
[the remark this answers is not in the document]
Comment (kadisi) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md]: yes

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+         * [NodePool API](#nodepool-api)
+         * [UnitedDeployment API](#uniteddeployment-api)
+      * [Implementation History](#implementation-history)
+
+
+# Proposal about nodepool and uniteddeployment  
+
+## Glossary
+
+NodePool:
```
[the remark this answers is not in the document]
Comment (kadisi) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md]: NodePool is a new CRD resource used to represent a group of nodes. Nodes under the NodePool have the same attributes, such as geography, operating system, CPU architecture, and so on.

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+    
+## Summary
+
+In the edge scenario, compute nodes have strong regional attributes. There may only be one worker node in the same physical location, and there may also be multiple worker nodes in the same physical location. Therefore, from the work node resource perspective of edge nodes, they need to be divided into different node pools (NodePool) to represent the same set of features. After dividing nodes pool, there will be a demand for application of grouping management, users need to be deployed application according to the nodepool, combined with the concept of nodepool, to deploy applications on different nodes in the pool, pool dimensions at the nodes to expansion of application, upgrade, such as operation, at the same time, network access is also carried out in accordance with the node pool dimensions of network communication.
```
[the remark this answers is not in the document]
Comment (kadisi) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md]: good idea

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+- NodePool
+    Users can quickly learn which node pools or regions are in their cluster. You can also quickly see which nodes are in the node pool.
+    User can uniformly type label, annotation, and TAINts on nodes under the node pool.
+    This simplifies the operation and maintenance management of nodes.
+    
+- UnitedDeployment
+    User can use a template to deploy the application in a different node pool. This template can support one of Kubernetes deployment, Daemonset, Statefulset.
+    
+### Goals
+- Define the api of NodePool
```
[the remark this answers is not in the document]
Comment (kadisi) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md]: ok

File: docs/enhacement/20201211-nodepool_uniteddeployment.md (hunk 8 above)
Comment (charleszheng44) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md]: Yes. will change to `UnreadyNodeNum`

File: docs/enhacement/20201211-nodepool_uniteddeployment.md (hunk 7 above)
Comment (charleszheng44) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md]: Yes. The `ReadyNodeNum` is more clear.

File: docs/enhacement/20201211-nodepool_uniteddeployment.md (hunk 9 above)
Comment (charleszheng44) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md]: Will change to `condition type`

File: docs/enhacement/20201211-nodepool_uniteddeployment.md [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+- What is the plan for implementing this feature?
+
+  Provide yurtunit-manager operator to manage NodePool and UnitedDeployment crd
+
+### NodePool API
+
+``` go
+// NodePoolSpec defines the desired state of NodePool
+type NodePoolSpec struct {
+    // The type of the nodepool
```
[the remark this answers is not in the document]
Comment (charleszheng44) [reply docs/enhacement/20201211-nodepool_uniteddeployment.md]: will change to NodePool.

**Review by Fei-Guo — COMMENTED** [review]
It seems that most of the review feedbacks have been addressed. More tunes can be done in the actual implementations.

**Pull request merged**